### ISSUE TITLE
Transpose scheduler merges IterDomains with different iteration types.

### ISSUE THREAD

**Issue by jjsjann123** [NVIDIA/Fuser#1659]
repro script
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1, 1], contiguity=[True, True, None], dtype=DataType.Float, is_cpu=False, stride_order=[2, 1, 0])
    T1 = fd.define_tensor(shape=[-1, 1, -1], contiguity=[True, None, True], dtype=DataType.Float, is_cpu=False, stride_order=[2, 1, 0])
    T2 = fd.ops.sum(T0, axes=[1], keepdim=False, dtype=DataType.Null)
    T3 = fd.ops.sum(T1, axes=[1], keepdim=False, dtype=DataType.Null)
    T4 = fd.ops.mul(T2, T3)
    fd.add_output(T4)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn((524288,), dtype=torch.float32, device='cuda:0').as_strided((1024, 512, 1), (512, 1, 1)),
    torch.randn((524288,), dtype=torch.float32, device='cuda:0').as_strided((1024, 1, 512), (512, 512, 1)),
]
fd.execute(inputs)
```

Running into issues:
```
Traceback (most recent call last):
  File "/opt/pytorch/nvfuser/nvfuser/__init__.py", line 137, in execute
    result = self._execute(
RuntimeError: Merging IterDomains requires that their iteration types match. Outer: iS69{32}, Inner: rS7{i1}
Exception raised from merge at /opt/pytorch/nvfuser/csrc/ir/nodes.cpp:2692 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7fd2d0278a43 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #1: nvfuser::IterDomain::merge(nvfuser::IterDomain*, nvfuser::IterDomain*, bool) + 0x39b (0x7fd2d04fd27b in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #2: nvfuser::TensorDomain::merge(int, int) + 0xc9 (0x7fd2d04fd3d9 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #3: nvfuser::TensorView::merge(int, int) + 0xdb (0x7fd2d07cffcb in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #4: nvfuser::scheduleTranspose(nvfuser::Fusion*, nvfuser::TransposeParams) + 0x109b (0x7fd2d071fb5b in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #5: nvfuser::TransposeScheduler::schedule(nvfuser::Fusion*) + 0xe0 (0x7fd2d0721db0 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #6: nvfuser::FusionKernelRuntime::compileKernel(nvfuser::KernelArgumentHolder const&, nvfuser::SegmentedGroup*) + 0x151 (0x7fd2d0533ab1 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #7: nvfuser::FusionKernelRuntime::compileFusionParallel(nvfuser::KernelArgumentHolder) + 0x41c (0x7fd2d0539b9c in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #8: nvfuser::FusionExecutorCache::runFusionWithInputs(c10::ArrayRef<c10::IValue> const&, std::optional<nvfuser::PrimDataType>, std::optional<signed char>) + 0xa43 (0x7fd2d0545523 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #9: nvfuser::python_frontend::FusionDefinition::execute(c10::ArrayRef<c10::IValue> const&, bool, bool, std::optional<signed char>) const + 0x16c (0x7fd2d0854a2c in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #10: <unknown function> + 0xf152e (0x7fd2d0ba252e in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #11: <unknown function> + 0x177488 (0x7fd2d0c28488 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
<omitting python frames>
frame #27: <unknown function> + 0x29d90 (0x7fd4f73bbd90 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #28: __libc_start_main + 0x80 (0x7fd4f73bbe40 in /usr/lib/x86_64-linux-gnu/libc.so.6)
```

**Comment by jjsjann123** [comment]
It's asserting here: https://github.com/NVIDIA/Fuser/blob/da7c4e97e3b90267a5465de856afc5e8d64fc56a/csrc/scheduler/transpose.cpp#L1239C1-L1239C26

roughly looking like it's the propagation that's failing the task.
printing out the fusion right before the assert (after global scheduling is done.

global reference looks like this:
```
T4_g[ iblockIdx.x33{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(i6, 32) ) ), 1) )}, iUS34{1}, iS29{32}, iS31{32} ]
 root domain : (iS26{i0}, iS27{i6})
 contiguity: t t
  Split: iS26{i0} by factor 32 -> iS30{( ceilDiv(i0, 32) )}, iS31{32}, start offset: 0, stop offset: 0
  Split: iS27{i6} by factor 32 -> iS28{( ceilDiv(i6, 32) )}, iS29{32}, start offset: 0, stop offset: 0
  Merge: iS30{( ceilDiv(i0, 32) )} and iS28{( ceilDiv(i6, 32) )} -> iS32{( ( ceilDiv(i0, 32) ) * ( ceilDiv(i6, 32) ) )}
  Split: iS32{( ( ceilDiv(i0, 32) ) * ( ceilDiv(i6, 32) ) )} by factor 1 -> iblockIdx.x33{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(i6, 32) ) ), 1) )}, iUS34{1}, start offset: 0, stop offset: 0
 leaf domain : (iblockIdx.x33{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(i6, 32) ) ), 1) )}, iUS34{1}, iS29{32}, iS31{32})
```
But propagated reference2 looks like this:
```
T2_g[ iS75{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(1, 32) ) ), 1) )}, iS76{1}, bS73{32}, iS71{32}, rS7{i1} ]
 root domain : (iS6{i0}, rS7{i1}, bS8{1})
 contiguity: t n n
  Split: iS6{i0} by factor 32 -> iS70{( ceilDiv(i0, 32) )}, iS71{32}, start offset: 0, stop offset: 0
  Split: bS8{1} by factor 32 -> bS72{( ceilDiv(1, 32) )}, bS73{32}, start offset: 0, stop offset: 0
  Merge: iS70{( ceilDiv(i0, 32) )} and bS72{( ceilDiv(1, 32) )} -> iS74{( ( ceilDiv(i0, 32) ) * ( ceilDiv(1, 32) ) )}
  Split: iS74{( ( ceilDiv(i0, 32) ) * ( ceilDiv(1, 32) ) )} by factor 1 -> iS75{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(1, 32) 
) ), 1) )}, iS76{1}, start offset: 0, stop offset: 0
 leaf domain : (iS75{( ceilDiv(( ( ceilDiv(i0, 32) ) * ( ceilDiv(1, 32) ) ), 1) )}, iS76{1}, bS73{32}, iS71{32}, rS7{i1})
```
For some reason the `rS7{i1}` is reordered to be the inner most dimension and it doesn't look right. Since we are expecting the tile to be inner most.


segmented fusion for the reference:
```
g{(transpose)
inputs:
T1_g[ iS13{i0}, bS4{1}, iS5{i6} ] float
T2_g[ iS6{i0}, rS7{i1}, bS8{1} ] float
outputs:
T4_g[ iS11{i0}, iS12{i6} ] float
 
 
T3_g[ iS14{i0}, iS10{i6} ]
   = squeeze( T1_g[ iS13{i0}, bS4{1}, iS5{i6} ] )
(1)
T4_g[ iS11{i0}, iS12{i6} ]
   = T2_g[ iS6{i0}, rS7{i1}, bS8{1} ]
   * T3_g[ iS14{i0}, iS10{i6} ];
(2)
}
```

**Comment by jjsjann123** [comment]
logging offline conversation with @naoyam :

Suggested that the reduction domain here shouldn't matter and we don't have a protocol on where to put that dangling reduction iterdomain on inputs.

I'll just re-order it past the tiling. :crossed_fingers: Hopefully this would work well with propagation to other tensors in the group2 for transpose scheduler.

**Comment by jjsjann123** [comment]
Note to myself. I'm also not totally sure how we are handling `dims_merged_with_1/2` in `maybeBuildVirtualInnerDims`. Are we safely ignoring reduction iterdomains there? seems not since we are naively looking at `shape_in_ref1`. I should try to tweak the repro to verify that as well.

**Comment by jjsjann123** [comment]
```TEST_F(TransposeTest, TrivialReductionIterDomainOnInputsIssueRepro1659_part2) {
  auto fusion = std::make_unique<Fusion>();
  auto fusion_ptr = fusion.get();
  FusionGuard fg(fusion_ptr);

  auto tv0 = TensorViewBuilder()
                 .ndims(4)
                 .contiguity({true, std::nullopt, true, std::nullopt})
                 .shape({-1, 1, -1, 1})
                 .dtype(DataType::Float)
                 .build();
  fusion->addInput(tv0);
  auto tv1 = TensorViewBuilder()
                 .ndims(4)
                 .contiguity({true, true, std::nullopt, true})
                 .shape({-1, -1, 1, -1})
                 .dtype(DataType::Float)
                 .build();
  fusion->addInput(tv1);
  auto tv2 = sum(tv0, {2});
  auto tv3 = squeeze(tv1, std::vector<int64_t>{2});
  auto tv4 = add(tv2, tv3);
  fusion->addOutput(tv4);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);

  auto t0 = at::randn({1024, 1, 512, 1}, options);
  auto t1 = at::randn({1024, 128, 1, 4}, options);
  std::vector<c10::IValue> aten_inputs({t0, t1});

  FusionExecutorCache executor_cache(std::move(fusion));
  auto cg_outputs = executor_cache.runFusionWithInputs(aten_inputs);

  auto runtime = executor_cache.getMostRecentKernelRuntime();
  NVF_CHECK(runtime->isSegmented(), "Segmentation expected");
  auto heuristic0 =
      runtime->schedulerHeuristics()->heuristicsList().at(0).get()->heuristic();
  NVF_CHECK(
      heuristic0 == ScheduleHeuristic::Reduction,
      "Unexpected heuristic: ",
      heuristic0);
  auto heuristic1 =
      runtime->schedulerHeuristics()->heuristicsList().at(1).get()->heuristic();
  NVF_CHECK(
      heuristic1 == ScheduleHeuristic::Transpose,
      "Unexpected heuristic: ",
      heuristic1);
  auto tv_ref = t0.sum(2) + t1.squeeze(2);
  testValidate(fusion_ptr, cg_outputs, {t0, t1}, __LINE__, __FILE__);
  // testValidate(fusion_ptr, cg_outputs, {t0, t1}, {tv_ref}, __LINE__, __FILE__);
}
```

This is a bit scary, I'm getting wrong result from this example above (after the fix in #1661). Let me figure out if that fix accidentally impacted merge axis. :crossed_fingers:

**Comment by jjsjann123** [comment]
I think the indexing issue is a separate thing with the failing example above. I'll open a separate issue to track that. Meanwhile, I think the reorder thing should be a small enough thing that I'd like it to go in as a standalone PR. Even though it exposed the other indexing issue (maybe?!)
